packages/backend/server/src/modules/features/types.ts

```diff
@@ -0,0 +1,33 @@
+import type { Prisma } from '@prisma/client';
+
+export enum FeatureKind {
+  Feature,
+  Quota,
+}
+
+export type CommonFeature = {
+  feature: string;
+  type: FeatureKind;
+  version: number;
+  configs: Prisma.InputJsonValue;
+};
+
+export type Feature = CommonFeature & {
+  type: FeatureKind.Feature;
+  feature: FeatureType;
+};
+
+export enum FeatureType {
+  Feature_EarlyAccess = 'early_access',
+}
+
+export const Features: Feature[] = [
+  {
+    feature: FeatureType.Feature_EarlyAccess,
+    type: FeatureKind.Feature,
+    version: 1,
+    configs: {
```

Comment:
The untyped configs seem hard to be used, like `whitelist: String[]` here.

We could make it a series of Feature/Quota checkers with serializable/deserializable capability.

```tsx
class EarlyAccessFeature<T extends User> implements Feature<T> {
  name = Features.EarlyAccess
  kind = FeatureKind.Feature
  
  constructor(config: string /* from db */) {
    this.configs = Config.deser(
      config, 
      ['whitelist'] as const /* only show known configs by ts */
    )
  }

  validate(user: T): Promise<boolean> {
    if (this.checkWhitelist(user.email)) {
      return true
    }

    // ... custom check
  }

  checkWhitelist(email: string) {
    // type error
    // this.configs.unknown.xxx
    return this.configs.whitelist.some(suffix => email.endsWith(suffix))
  }
}

class BlobQuota<T extends number> implements Quota<T> {
  name = Features.BlobQuota
  kind = FeatureKind.Quota

  consturctor(config: string) {
    this.config = Config.deser(config, ['blobLimit'] as const)
  }

  next(_curr: T, next: T): Promise<boolean> {
    return this.config.blobLimit >= nextVal
  }
}

class StorageQuota<T extends number> implements Quota<T> {
  name = Features.StorageQuota
  kind = FeatureKind.Quota

  consturctor(config: string) {
    this.configs = Config.deser(config, ['storageQuota'] as const)
  }

  next(curr: T, next: T): Promise<boolean> {
    return this.configs.storageQuota >= curr + next
  }
}

const configsMap = {
  whitelist: WhitelistConfig,
  blobLimit: NumberConfig,
  storageLimit: AccumulationConfig
  // ...
}

class ConfigFactor {
  static deser<T extends string[]>(str: string, knownConfigFields: T): { [key in T[number]: Config } {
    const configs = Object.entries(JSON.parse(str))
			.forEach(([key, val]) => {
        configs[key] = ConfigFactor.create(key, val)
      })
  }

  ser(configs: Record<string, Config>) {
    return JSON.stringify(configs)
  }

  static create(name: string, c: any) {
    return new configsMap[name](c)
  }
}
```

Reply:
runtime check for config has been impl at #5142